### ISSUE TITLE
Fix JSON data content type mess

### ISSUE THREAD

**Issue by jskeet** [cloudevents/sdk-csharp#200]
It looks like I made a mistake when it comes to data content type defaulting in recent changes. I'd raised [an issue in the spec repo](https://github.com/cloudevents/spec/issues/933) to get clarification - but then failed to act on the resolution.

To cut a long story short, in the JSON event format, we need to work out what to do when we are serializing an event and have data but no datacontenttype attribute. The spec says (and will be clarified further to say):

- For binary data, don't infer anything
- For anything else, infer a data content type of "application/json" and proceed as if it had been specified explicitly

In our current implementation (for both Json.NET and System.Text.Json):

- When serializing in structured mode, we infer a data content type of "application/json" for *any* event with a non-null Data property but a null DataContentType property
- When serializing in binary mode, an event format has no way of advertising a "default" data content type, so when using HTTP transport for example, the content type header *isn't* populated.

Three possible approaches. Personally I favor option 3, but would value more input.

## Option 1: add data content type defaulting to EncodeStructuredData responsibilities

In this option, we'd move the "if we think we're dealing with JSON and don't have a data content type, emit the `datacontenttype` property with "application/json"" code into EncodeStructuredData, and change the documentation to say that an implementation *can* emit that property.

Pros:

- Solves the structured mode default implementation issue

Cons:

- Changes the contract for implementations overriding EncodeStructuredData somewhat unpleasantly
- Doesn't address the binary mode issue

## Option 2: remove the "make datacontenttype explicit" code entirely

In this option, we'd just remove the code that populates the `datacontenttype` property in structured mode. This is probably okay by the letter of the spec, but feels like it goes against the spirit.

Pros:

- Simple and doesn't affect derived implementations
- Consistent between structured and binary mode

Cons:

- Moves further away from the spirit of the spec, for both structured and binary mode

## Option 3: add a virtual method in EventFormatter

This would be something like:

```csharp
public virtual ContentType? InferDataContentType(object data) => null;
```

Possibly variations:

- Could return a `string?` instead of `ContentType?`
- Could accept the `CloudEvent` instead of just the data

Our JSON event formatters would then override this to return "application/json" only in the case where the content isn't binary. Our various transport implementations could call this method when a `CloudEvent.DataContentType` is null, and so could the JSON event formatter WriteCloudEventForBatchOrStructuredMode methods.

Pros:

- Makes both binary and structured content modes consistent with the spec
- Expresses the idea of data content type defaulting in a single place (this has already been discussed a little in #190 as a consistency problem) which allows other event formats to do likewise
  - This could be regarded as a "con", if we think that the JSON event format effectively made a mistake by including the idea of defaulting.

Cons:

- Modifies the contract for derived JSON event formatters, in that any event formatter that overrides EncodeStructuredModeData and EncodeBinaryModeEventData should at least *consider* overriding this method as well

**Comment by jskeet** [comment]
I'll probably prototype option 3 today - I'm happy to prototype the other options as well for comparison if that would be useful.

**Comment by jskeet** [comment]
#201 is a prototype. I'm reasonably happy with how it's worked out, and still think it's the best approach.

**Comment by JoshLove-msft** [comment]
Personally, I think Option 2 makes the most sense, as it seems beyond the scope of the SDK to set the datacontenttype on the user's behalf. I don't really see how this is going against the spirit of the spec - I don't see it mentioned anywhere that implementers are responsible for setting the content type, but rather it explains how to handle situations where the content type is not set. We can make inferences as per the spec, but actually modifying the data seems unnecessary. 

On the other hand, I could see how this information would be useful from an auditing perspective to understand how the SDK actually serialized the data. But still, I would lean towards just leaving the content type as-is. If you feel strongly about going ahead with your prototype, I think it is probably okay, but just wanted to provide my perspective.

**Comment by jskeet** [comment]
Perspective is good :)

The reason I think that just reading the spec would lead to a `datacontenttype` of application/json is:

> If the datacontenttype is unspecified, processing SHOULD proceed as if the datacontenttype had been specified explicitly as `application/json`

... it depends on what we mean by "processing" here. If it's "processing just the data" then it's fine to miss it off. If it's "processing the whole event" then there's a difference in terms of whether the `datacontenttype` is included in the structured mode representation.

On the one hand, the example of "Example event with a literal JSON string as the non-Binary-valued data and no datacontenttype" *doesn't* show a `datacontenttype` in the structured mode... but it *does* show it in binary mode, and there's this in the spec:

> Note that the Content Type is explicitly set to the application/json value that was implicit in JSON format.

One new option would be a hybrid of options 2 and 3 - all event formatters to infer the content type for *binary* mode, but don't use that within structured mode...

**Comment by JoshLove-msft** [comment]
Fair points - I think the fact that a user is opting into the usage of the various formatters, setting the datacontenttype on the serialized payload is probably reasonable. We aren't modifying the `CloudEvent.DataContentType` I don't think, which is what I was thinking initially.

**Comment by jskeet** [comment]
Right - I *definitely* don't want to modify the CloudEvent that's passed into us. Apologies for the confusion on that front.

So do you have any preference between option 3 (set the `datacontenttype` in structured mode and `content-type` in binary mode) or the new option 4 (leave it implicit in structured mode but set `content-type` in binary mode)?

**Comment by JoshLove-msft** [comment]
In terms of the hybrid approach, it might be more explainable if we just set the content type for both modes. It seems like a useful convenience for users, and I don't really see a reason not to set it when a user opts in to a specific formatter, and hasn't set a content type themselves.

**Comment by jskeet** [comment]
> In terms of the hybrid approach, it might be more explainable if we just set the content type for both modes. It seems like a useful convenience for users, and I don't really see a reason not to set it when a user opts in to a specific formatter, and hasn't set a content type themselves.

Agreed. In that case, I'll fill out the rest of the prototype, if that makes sense to you. Please feel free to comment on the prototype PR in terms of naming etc - although it's easy enough to change the names shortly before merging too, so please don't feel any pressure. Thanks for the swift discussion on this - most useful.

**Comment by jskeet** [comment]
Fixed in #201.
